### ISSUE TITLE
Using proxy_server and/or proxy_port has no effect on Windows

### ISSUE THREAD

**Issue by TeunRoefs** [voxpupuli/puppet-archive#277]
## Affected Puppet, Ruby, OS and module versions/distributions

- Puppet: 4.8.1
- Ruby: 1.9.3
- Distribution: -
- Module version: 1.3.0

## How to reproduce (e.g Puppet code you use)
```
archive { "C:\\Windows\\Temp\\SysinternalsSuite.zip":
      ensure       => present,
      source       => 'https://live.sysinternals.com/files/SysinternalsSuite.zip',
      proxy_server => 'http://1.2.3.4:3120', #use own proxy here
      cleanup      => false,
}
```
## What are you seeing
Absolutely no traffic is going through our proxy in regards to the above configuration. The exact configuration is working on Linux (Debian/RedHat). 

File is download without going through proxy.

## What behaviour did you expect instead
A CONNECT log event connecting to live.sysinternals.com that allowed the download and archiving of the SysinteralsSuite.zip file.

## Any additional information you'd like to impart
If there are any more questions I am happy to provide extra information.

**Comment by alexjfisher** [comment]
Can you try setting `proxy_type => 'https'`?

This line https://github.com/voxpupuli/puppet-archive/blob/36e1a9bd90d5fd36df2a526983ddc3b82c06392a/lib/puppet/type/archive.rb#L260 looks buggy to me.  It defaults `proxy_type` based on `proxy_server` instead of `source`.

**Comment by TeunRoefs** [comment]
Hi Alex,

Thank you for the fast response.
Either using ``proxy_type => 'https'`` or ``proxy_type => 'http'`` has no effect on the outcome.

**Comment by nanliu** [comment]
So on windows this uses the ruby provider, and per stdlib, it's suppose to honor the [environment settings](https://ruby-doc.org/stdlib-2.4.1/libdoc/net/http/rdoc/Net/HTTP.html#class-Net::HTTP-label-Proxies):

> Net::HTTP will automatically create a proxy from the http_proxy environment variable if it is present. To disable use of http_proxy, pass nil for the proxy address.

Please add a few `Puppet.debug` lines to check and see:
* is the environment var set correctly? (is it set correctly [here](https://github.com/voxpupuli/puppet-archive/blob/master/lib/puppet_x/bodeco/util.rb#L31)
* does net::http honor this env var? (is still set correctly [here](https://github.com/voxpupuli/puppet-archive/blob/master/lib/puppet_x/bodeco/util.rb#L51)

This is most likely windows specific and we may need to switch to the explicit proxy declaration instead depending on the ENV var behavior.

**Comment by alexjfisher** [comment]
Earlier today, I tried the ruby provider out on Linux and also observed the proxy settings didn't work.  This was for an http url with an http proxy (no https anywhere).  Not sure at what point the environment variable stopped working (I'm assuming it did work at some point!).  I tried puppet 3 on ruby 2.1 and puppet 4 on 2.3 IIRC.

**Comment by TeunRoefs** [comment]
What is the next step? Is there anything you want me to do in regards to this issue. Also, the remark was pointed at me @nanliu ?

**Comment by nanliu** [comment]
@TeunRoefs, based on documentation, I replace ENV with specific flags in PR #280. I wanted to provided a quick test patch, so I didn't get a chance to test locally. Let me know if that fixes this issue.

**Comment by TeunRoefs** [comment]
@nanliu Thank you for your update. I am getting this error when testing it locally.

```
Error: Could not set 'present' on ensure: no implicit conversion of Hash into String at C:/Users/squid/Downloads/staging
.pp:1
Error: Could not set 'present' on ensure: no implicit conversion of Hash into String at C:/Users/squid/Downloads/staging
.pp:1
Wrapped exception:
no implicit conversion of Hash into String
Error: /Stage[main]/Main/Archive[C:\Windows\Temp\SysinternalsSuite.zip]/ensure: change from absent to present failed: Co
uld not set 'present' on ensure: no implicit conversion of Hash into String at C:/Users/squid/Downloads/staging.pp:1
```

**Comment by TeunRoefs** [comment]
@nanliu Put a comment on your PR on the line that needs to be fixed. After changing the way the extra parameters are called in Net::HTTP.start() the Proxy server is working.

**Comment by TeunRoefs** [comment]
@alexjfisher What is the release sequence of the puppet-archive module? Kinda need the functionality in PR #280 created by @nanliu Is there anything you guys want me to do?

**Comment by krjensen** [comment]
Tested - works fine. Could we get this merged to master please?

**Comment by nanliu** [comment]
Closed per #280.
